client/src/main/java/io/pravega/client/stream/impl/ReaderGroupStateManager.java

```diff
@@ -421,4 +414,10 @@ void checkpoint(String checkpointName, PositionInternal lastPosition) throws Rei
             throw new ReinitializationRequiredException();
         }
     }
+
+    public String refreshDelegationToken(Segment segmentId) {
+        synchronized (this) {
+            return getAndHandleExceptions(controller.getOrRefreshDelegationTokenFor(segmentId.getScope(), segmentId.getStreamName()), RuntimeException::new);
```

Comment:
You are waiting of a future while holding a lock. This is bad multithreading practice.

Reply:
Fixed. We do not need lock as we are not changing any global state.